src/pc/common/BitArray.js

```diff
@@ -1,143 +1,73 @@
-// const assert = require('assert')
-const neededBits = require('./neededBits')
-
 class BitArray {
-  constructor (options) {
-    if (options === null) {
-      return
-    }
-    // assert(options.bitsPerValue > 0, 'bits per value must at least 1')
-    // assert(options.bitsPerValue <= 32, 'bits per value exceeds 32')
-
-    const length = Math.ceil((options.capacity * options.bitsPerValue) / 64)
-    if (!options.data) {
-      options.data = Array(length * 2).fill(0)
-    }
-    const valueMask = (1 << options.bitsPerValue) - 1
-
-    this.data = options.data
-    this.capacity = options.capacity
-    this.bitsPerValue = options.bitsPerValue
-    this.valueMask = valueMask
-  }
-
-  toJson () {
-    return JSON.stringify({
-      data: this.data,
-      capacity: this.capacity,
-      bitsPerValue: this.bitsPerValue,
-      valueMask: this.valueMask
-    })
-  }
-
-  static fromJson (j) {
-    const parsed = JSON.parse(j)
-    const bitarray = new BitArray(null)
-    bitarray.data = parsed.data
-    bitarray.capacity = parsed.capacity
-    bitarray.bitsPerValue = parsed.bitsPerValue
-    bitarray.valueMask = parsed.valueMask
-    return bitarray
+  constructor ({ bitsPerValue, capacity, data } = {}) {
+    this.bitsPerValue = bitsPerValue >>> 0
+    this.capacity = capacity >>> 0
+    this.data = data
+      ? (data.buffer ? new Uint32Array(data.buffer) : Uint32Array.from(data))
+      : new Uint32Array((this.capacity * this.bitsPerValue + 31) >>> 5)
```

Comment:
This is wrong: if capacity and bitsPerValue are both odd, the length of the array will be odd too (which cannot be as its supposed to represent an array of Long (64bits)).

It is working for our current use case as we only use power of 2 capacity. But this may break in the future if mojang decide to do weird things (again). The computation we had before this PR was giving the correct result, and I very much doubt this change is bringing any performance improvement.

Reply:
Okay, I'll revert it